034-kafka-topic-replication-factor.md

```diff
@@ -0,0 +1,76 @@
+
+The goal of this proposal is to offer users a simple approach to change the topic RF through the KafkaTopic Custom Resources and Cruise Control, in the cloud native, strimzi style. 
+
+## Current situation
+
+Currently, users have two options to change the topic RF:
+* Leverage the Native Kafka script to change the RF. Users need to generate the partition replica assignment and balance the kafka clusters manually if needed. 
+* Delete the topic and recreate the topic with the desired RF.
+
+Clearly none of the above approaches is user-friendly.
+
+When users attempt to change the RF in the KafkaTopic Resources, the Status of the KafkaTopic will become `NotReady` with `ReplicationFactorChangeException`. The following error message is shared with users.
+
+`Changing 'spec.replicas' is not supported. This KafkaTopic's 'spec.replicas' should be reverted to 4 and then the replication should be changed directly in Kafka.`
+
+## Motivation
+
+With this new feature, it will become much easier to achieve the desired tradeoff between the cost and data reliability in the kafka clusters. 
+
+For instance, users can lower the RFs for selective less critical topics to reduce the infrastructure cost (Compute, Storage and Network). Also, for some critical kafka topics, users can increase the data reliability through the bump of the RFs ONLY for these specific topics. 
+
+More importantly, users only need to specify the new RF number in the `KafkaTopic` resource and approve the change, in order to change the RF for kafka topics. The heavy lifting work, such as new replica placement, extra replica deletion and possible cluster rebalance, is completely taken care by the strimzi. 
+
+## Proposal
+
+### API
+The Kafka Topic RF can be changed by users in the following steps:
+* Update the `spec.replicas` properties in the target `KafkaTopic` Resource in the kubernetes cluster
+* Watch the `Status` field of the `KafkaTopic Resource`, for the status update of this RF change.
+
+### Design
+We propose to change the topic RFs with the support of `Cruise Control` and `Strimzi Topic Operator`. 
+
+#### Cruise Control
+Cruise Control Service will be leveraged to change the RF for topics in the kafka clusters, for the following reasons: 
+* Strimzi supports the Cruise Control Service, and it's ready for use with simple configurations. 
+* Cruise Control has native support for the RF changes. It automates the whole process, including selecting of new replicas and extra replicas, triggering the replica creation in the kafka brokers and monitoring the cluster during the process.
+
+#### Kafka Topic Operator
+Strimzi Topic Operator will watch the changes of `KafkaTopic` Resource. 
+
+For any new topic RF change event, the `Strimzi Topic Operator` will validate the change. For instance, the new RF should be >= the min insync replica set size for the target topic. 
```

Comment:
Is this the only validation condition, or are there others? I guess we can't have > #brokers. But perhaps the TO should support a separate limit that's less than that, to allow someone administering the Kafka cluster to place an upper bound for people changing the `KafkaTopic` (that might be different people).

Reply:
I added the #brokers configuration in the proposal.

For the upper bound of the RF, what's the reasonable number from your perspective? I feel 5 should be enough, but I could be wrong.